### PR TITLE
Add ESP readout support to Qiskit

### PR DESCRIPTION
**Pull request by zachschoenfeld33**
### Summary

Adds flag `use_measure_esp` to `assemble`, which sets the final measurement in a circuit to use Excited State Promoted (ESP) readout. See https://arxiv.org/pdf/2008.08571.pdf for reference.

A backend configuration flag `measure_esp_enabled` is included. The ESP feature can only be used if `measure_esp_enabled=True`.

### CODE REVIEW

**Comment by zachschoenfeld33** [comment]
@taalexander @dtmcclure

File: qiskit/compiler/assembler.py [hunk 1]
```diff
@@ -140,7 +141,11 @@ def assemble(
 
             ['gaussian', 'constant']
         init_qubits: Whether to reset the qubits to the ground state for each shot.
-                     Default: ``True``.
+            Default: ``True``.
+        use_measure_esp: Whether to use ESP (excited state promoted) readout for the final
```
Comment (taalexander) [inline qiskit/compiler/assembler.py:145]:
```suggestion
        use_measure_esp: Whether to use excited state promoted (ESP) readout for the final
```

File: qiskit/compiler/assembler.py [hunk 2]
```diff
@@ -140,7 +141,11 @@ def assemble(
...
+            measurement in each circuit. ESP readout can offer higher fidelity than standard
+            measurement sequences. See `here <https://arxiv.org/pdf/2008.08571.pdf>`_.
+            Default (set on backend): ``True`` if backend supports ESP readout, else ``False``.
```
Comment (taalexander) [inline qiskit/compiler/assembler.py:148]:
```suggestion
            Defaults to ``True`` if the backend supports ESP readout, else ``False``.
```

File: qiskit/compiler/assembler.py [hunk 3]
```diff
@@ -343,13 +352,21 @@ def _parse_common_args(
         for lo_config in schedule_los
     ]
 
+    measure_esp_enabled = getattr(backend_config, "measure_esp_enabled", False)
+    if use_measure_esp is None:
+        use_measure_esp = True if measure_esp_enabled else False  # default use of esp readout
+    if not measure_esp_enabled and use_measure_esp:
+        raise QiskitError("ESP readout not supported on this device. Please make sure the flag "
+                          "'use_measure_esp' is set to 'False' or not used.")
```
Comment (taalexander) [inline qiskit/compiler/assembler.py:360]:
```suggestion
                          "'use_measure_esp' is unset or set to 'False'.")
```

File: qiskit/execute_function.py [hunk 4]
```diff
@@ -69,6 +69,7 @@ def execute(
     meas_map=None,
     scheduling_method=None,
     init_qubits=None,
+    use_measure_esp=None,
```
Comment (taalexander) [inline qiskit/execute_function.py:72]: Is it possible to set this as a run_config argument, or must this be passed explicitly to `assemble`? Perhaps we could ask for this to be set in `run_config` and pulse this out to place it in `assemble`? This way we keep the `execute` signature generic to any backend. In the long run `assemble` will be moved into the compiler so changes there are Ok.

File: qiskit/providers/models/backendconfiguration.py [hunk 5]
```diff
@@ -300,6 +301,10 @@ def __init__(
                 - segment: Segment this processor belongs to within a larger chip.
             parametric_pulses (list): A list of pulse shapes which are supported on the backend.
                 For example: ``['gaussian', 'constant']``
+            measure_esp_enabled (bool): Whether ESP readout is enabled on this device. ESP readout
```
Comment (taalexander) [inline qiskit/providers/models/backendconfiguration.py:304]:
```suggestion
            measure_esp_enabled (bool): Whether excited state promoted (ESP) readout is enabled on this device. ESP readout
```

File: qiskit/schemas/backend_configuration_schema.json [hunk 6]
```diff
@@ -157,6 +157,11 @@
                     "description": "Whether delay between programs can be set dynamically using 'rep_delay').",
                     "default": false
                     },
+                "measure_esp_enabled": {
```
Comment (taalexander) [inline qiskit/schemas/backend_configuration_schema.json:160]: Version bump here as well.

**Review by taalexander — CHANGES_REQUESTED** [review]

File: qiskit/schemas/backend_configuration_schema.json (hunk 6 above)
Comment (zachschoenfeld33) [reply qiskit/schemas/backend_configuration_schema.json:160]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/6543/commits/de6d6073315d5607ef59c929861b38aa6d47847a.

File: qiskit/providers/models/backendconfiguration.py (hunk 5 above)
Comment (zachschoenfeld33) [reply qiskit/providers/models/backendconfiguration.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/6543/commits/de6d6073315d5607ef59c929861b38aa6d47847a.

File: qiskit/compiler/assembler.py (hunk 3 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assembler.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/6543/commits/de6d6073315d5607ef59c929861b38aa6d47847a.

File: qiskit/compiler/assembler.py (hunk 2 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assembler.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/6543/commits/de6d6073315d5607ef59c929861b38aa6d47847a

File: qiskit/compiler/assembler.py (hunk 1 above)
Comment (zachschoenfeld33) [reply qiskit/compiler/assembler.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/6543/commits/de6d6073315d5607ef59c929861b38aa6d47847a.

File: qiskit/execute_function.py (hunk 4 above)
Comment (zachschoenfeld33) [reply qiskit/execute_function.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/6543/commits/9af1534eb4eb37589a5ecef10a0521ff0d3ecc86.

**Review by taalexander — APPROVED** [review]

**Comment by mtreinish** [comment]
I'm not clear why this needed to be merged into terra at all. It's an ibm backend only option right? Why isn't this just exposed as an option from the ibmq provider backend? It should be passed through the kwargs for the run command (and execute() by extension) if it's exposed through the options interface correctly.

Also we shouldn't be updating the schemas here anymore. They're deprecated and frozen slated for removal soon (see #6558 )

**Comment by taalexander** [comment]
> I'm not clear why this needed to be merged into terra at all. It's an ibm backend only option right? Why isn't this just exposed as an option from the ibmq provider backend? It should be passed through the kwargs for the run command (and execute() by extension) if it's exposed through the options interface correctly.

Agreed, this is a backend-only option. Unfortunately, the provider is calling out to `assemble` to build the Qobj payload still, so until this function is migrated to the provider this seems like the required place for this logic. We avoided any other IBM specific changes in this PR. See https://github.com/Qiskit/qiskit-ibmq-provider/pull/952 for backend-specific changes. 

> Also we shouldn't be updating the schemas here anymore. They're deprecated and frozen slated for removal soon (see #6558 )
 
 Ok, these changes are just a mirror of https://github.com/Qiskit/ibm-quantum-schemas/pull/10. We will not make further updates to these schemas.

**Comment by mtreinish** [comment]
> > I'm not clear why this needed to be merged into terra at all. It's an ibm backend only option right? Why isn't this just exposed as an option from the ibmq provider backend? It should be passed through the kwargs for the run command (and execute() by extension) if it's exposed through the options interface correctly.
> 
> Agreed, this is a backend-only option. Unfortunately, the provider is calling out to `assemble` to build the Qobj payload still, so until this function is migrated to the provider this seems like the required place for this logic. We avoided any other IBM specific changes in this PR. See [Qiskit/qiskit-ibmq-provider#952](https://github.com/Qiskit/qiskit-ibmq-provider/pull/952) for backend-specific changes.

Sure, I agree that we should copy the assembler and `qiskit.qobj` over to the ibmq provider and longer term deprecate and remove it from terra (but we should be conservative on that given it's former central usage). But, even without that being done the way `assemble()` is written you don't need to add explicit kwargs for every single option in the run config. If you pass in arbitrary kwargs to the function it gets added directly the run config. For example I ran:

```python3
from qiskit.compiler import assemble
from qiskit import QuantumCircuit
import qiskit

print(qiskit.__version__)
qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)
qc.measure_all()
print(assemble(qc, use_measure_esp=True))
```

with the current release of qiskit-terra and it returns:

```
0.17.4
QASM Qobj: a5005853-6158-4e00-9abe-9f061adeb129:
Config: {'init_qubits': True,
 'meas_level': <MeasLevel.CLASSIFIED: 2>,
 'memory': False,
 'memory_slots': 2,
 'n_qubits': 2,
 'parameter_binds': [],
 'shots': 1024,
 'use_measure_esp': True}
Header: {}
Experiments:

QASM Experiment:
Header:
{'clbit_labels': [['meas', 0], ['meas', 1]],
 'creg_sizes': [['meas', 2]],
 'global_phase': 0.0,
 'memory_slots': 2,
 'metadata': {},
 'n_qubits': 2,
 'name': 'circuit-0',
 'qreg_sizes': [['q', 2]],
 'qubit_labels': [['q', 0], ['q', 1]]}
Config:
{'memory_slots': 2, 'n_qubits': 2}

	Instruction: h
		qubits: [0]

	Instruction: cx
		qubits: [0, 1]

	Instruction: barrier
		qubits: [0, 1]

	Instruction: measure
		qubits: [0]
		memory: [0]

	Instruction: measure
		qubits: [1]
		memory: [1]
```

So all this is doing in practice is raising an exception at `assemble()` call time if `use_measure_esp` is enabled and the backend doesn't support it. But that arguably should be checked before the provider is calling to terra's `assemble()` function.

**Comment by zachschoenfeld33** [comment]
One thing I will say is it was far easier to debug and test from assemble than the provider. `assemble` has local unit tests, while the provider seems to require some server connections...

**Comment by zachschoenfeld33** [comment]
This can be revoked in favor of https://github.com/Qiskit/qiskit-ibmq-provider/pull/952.

**Pull request merged**